tests/unittests/algo/test_tpe.py

```diff
@@ -0,0 +1,220 @@
+
+
+@pytest.fixture()
+def space():
+    """Return an optimization space"""
+    space = Space()
+    dim1 = Real('yolo1', 'uniform', -3, 6)
+    space.register(dim1)
+    dim2 = Real('yolo2', 'uniform', -2, 4)
+    space.register(dim2)
+
+    return space
+
+
+@pytest.fixture
+def tpe(space):
+    """Return an instance of TPE."""
+    return TPE(space)
+
+
+@pytest.fixture
+def gmm_sampler(tpe):
+    """Return an instance of GMMSampler"""
+    mus = numpy.linspace(-3, 3, num=12, endpoint=False)
+    sigmas = [0.5] * 12
+
+    return GMMSampler(tpe, mus, sigmas, -3, 3)
+
+
+def test_compute_max_ei_point():
+    """Test that max ei point is computed correctly"""
+    points = numpy.linspace(-3, 3, num=10)
+    below_likelis = numpy.linspace(0.5, 0.9, num=10)
+    above_likes = numpy.linspace(0.2, 0.5, num=10)
+    max_ei_point = compute_max_ei_point(points, below_likelis, above_likes)
+    assert max_ei_point == 3
+
+
+def test_adaptive_parzen_normal_estimator():
+    """Test adaptive kernel estimator"""
+    obs_mus = numpy.linspace(-1, 5, num=30, endpoint=False)
+    low = -1
+    high = 5
+    mus, sigmas, weights = adaptive_parzen_estimator(obs_mus, low, high, prior_weight=1.0,
+                                                     equal_weight=False, flat_num=25)
+
+    assert len(mus) == len(sigmas) == len(weights) == 30 + 1
+    assert numpy.all(weights[-25:] == weights[-1])
+    assert numpy.all(sigmas == 6 / 10)
+
+    # equal weight
+    mus, sigmas, weights = adaptive_parzen_estimator(obs_mus, low, high, prior_weight=1.0,
+                                                     equal_weight=True, flat_num=25)
+    assert numpy.all(weights[-25:] == weights[-1])
+    assert numpy.all(sigmas == 6 / 10)
+
+    # priori weight
+    mus, sigmas, weights = adaptive_parzen_estimator(obs_mus, low, high, prior_weight=0.5,
+                                                     equal_weight=False, flat_num=25)
+    priori_pos = numpy.searchsorted(mus, 2)
+    assert weights[priori_pos] == 0.5 / 31
+    assert numpy.all(weights[31 - priori_pos:] == weights[-1])
+    assert numpy.all(sigmas == 6 / 10)
+
+    # sigma clip
+    obs_mus = numpy.linspace(-1, 5, num=8, endpoint=False)
+    mus, sigmas, weights = adaptive_parzen_estimator(obs_mus, low, high, prior_weight=1.0,
+                                                     equal_weight=False, flat_num=25)
+    assert numpy.all(sigmas == 6 / 8)
+
+
+class TestGMMSampler():
+    """Tests for TPE GMM Sampler"""
+
+    def test_gmm_sampler_creation(self, gmm_sampler):
+        """Test GMMSampler creation"""
+        assert len(gmm_sampler.weights) == 12
+        assert len(gmm_sampler.pdfs) == 12
+
+    def test_sample(self, gmm_sampler):
+        """Test GMMSampler sample function"""
+        points = gmm_sampler.sample(25)
+        points = numpy.array(points)
+
+        assert len(points) <= 25
+        assert numpy.all(points >= -3)
+        assert numpy.all(points < 3)
```

Comment:
We should also add a test_weight which samples many points and verify that the samples are properly spread according to weights. This could be done with very different mus and small sigmas.